### DISCUSSION TITLE
Input not working

### DISCUSSION THREAD

**Discussion by quodix** [Pi4J/pi4j#291] [Q&A]
Good evening everybody.
I have a problem, that my cable bridge (simulates a button) is not regognized by the code.

That is my current code, that is from the Pi4J example:
`import com.pi4j.Pi4J;
import com.pi4j.io.gpio.digital.DigitalInput;
import com.pi4j.io.gpio.digital.DigitalInputProvider;
import com.pi4j.io.gpio.digital.PullResistance;
import com.pi4j.util.Console;

public class TestMain {
    /**
     * Constant <code>DIGITAL_INPUT_PIN=0</code>
     */
    public static final int DIGITAL_INPUT_PIN = 21;

    /**
     * <p>Constructor for DigitalInputExample.</p>
     */
    public TestMain() {
    }

    /**
     * <p>main.</p>
     *
     * @param args an array of {@link java.lang.String} objects.
     * @throws java.lang.Exception if any.
     */
    public static void main(String[] args) throws Exception {
        // configure logging output
        System.setProperty("org.slf4j.simpleLogger.defaultLogLevel", "INFO");

        // TODO :: REMOVE TEMPORARY PROPERTIES WHEN NATIVE PIGPIO LIB IS READY
        // this temporary property is used to tell
        // PIGPIO which remote Raspberry Pi to connect to
        System.setProperty("pi4j.host", "rpi3bp.savage.lan");

        // create Pi4J console wrapper/helper
        // (This is a utility class to abstract some of the boilerplate stdin/stdout code)
        final var console = new Console();

        // print program title/header
        console.title("<-- The Pi4J Project -->", "Basic Digital Input Example");

        // allow for user to exit program using CTRL-C
        console.promptForExit();

        // Initialize Pi4J with an auto context
        // An auto context includes AUTO-DETECT BINDINGS enabled
        // which will load all detected Pi4J extension libraries
        // (Platforms and Providers) in the class path
        var pi4j = Pi4J.newAutoContext();

        // create a digital input instance using the default digital input provider
        // we will use the PULL_DOWN argument to set the pin pull-down resistance on this GPIO pin
        var config = DigitalInput.newConfigBuilder(pi4j)
                //.id("my-digital-input")
                .address(DIGITAL_INPUT_PIN)
                .pull(PullResistance.PULL_DOWN)
                .build();

        // get a Digital Input I/O provider from the Pi4J context
        DigitalInputProvider digitalInputProvider = pi4j.provider("pigpio-digital-input");

        var input = digitalInputProvider.create(config);

        // setup a digital output listener to listen for any state changes on the digital input
        input.addListener(event -> {
            Integer count = (Integer) event.source().metadata().get("count").value();
            console.println(event + " === " + count);
        });

        // lets read the analog output state
        console.print("THE STARTING DIGITAL INPUT STATE IS [");
        console.println(input.state() + "]");

        console.println("CHANGE INPUT STATES VIA I/O HARDWARE AND CHANGE EVENTS WILL BE PRINTED BELOW:");

        // wait (block) for user to exit program using CTRL-C
        console.waitForExit();

        // shutdown Pi4J
        console.println("ATTEMPTING TO SHUTDOWN/TERMINATE THIS PROGRAM");
        pi4j.shutdown();
    }

}`

The Console prints only this: 

`[main] INFO com.pi4j.util.Console - ************************************************************
[main] INFO com.pi4j.util.Console - ************************************************************
[main] INFO com.pi4j.util.Console -
[main] INFO com.pi4j.util.Console -                   <-- The Pi4J Project -->
[main] INFO com.pi4j.util.Console -                  Basic Digital Input Example
[main] INFO com.pi4j.util.Console -
[main] INFO com.pi4j.util.Console - ************************************************************
[main] INFO com.pi4j.util.Console - ************************************************************
[main] INFO com.pi4j.util.Console -
[main] INFO com.pi4j.util.Console - ------------------------------
[main] INFO com.pi4j.util.Console - |    PRESS CTRL-C TO EXIT    |
[main] INFO com.pi4j.util.Console - ------------------------------
[main] INFO com.pi4j.util.Console -
[main] INFO com.pi4j.Pi4J - New auto context
[main] INFO com.pi4j.Pi4J - New context builder
[main] INFO com.pi4j.platform.impl.DefaultRuntimePlatforms - adding platform to managed platform map [id=raspberrypi; name=RaspberryPi Platform; priority=5; class=com.pi4j.plugin.raspberrypi.platform.RaspberryPiPlatform]
[main] INFO com.pi4j.util.Console - THE STARTING DIGITAL INPUT STATE IS [
[main] INFO com.pi4j.util.Console - LOW]
[main] INFO com.pi4j.util.Console - CHANGE INPUT STATES VIA I/O HARDWARE AND CHANGE EVENTS WILL BE PRINTED BELOW:

`

Is the digital input pin the same as it is defined (21)? After a long search I found that is a "GPIO" Pin and not the real pin. GPIO-21 = PIN-29? But nothing happen if I short the cable to ground

**Comment by FDelporte** [comment]
BCM/GPIO 21 is actually pin 40 ...
https://pinout.xyz/pinout/pin40_gpio21/#

And pin 29 is BCM/GPIO 5
https://pinout.xyz/pinout/pin29_gpio5/#

yes, confusing, I know ;-)

you can also check this example that you can run easily with JBang, and the wiring diagram shows you the BCM/GPIO numbering of the Raspberry Pi header
https://pi4j.com/examples/jbang/jbang_minimal_example/

![](https://pi4j.com/assets/getting-started/minimal/led-button_bb.png)

**Reply by quodix** [reply]
Do I need JBang or can I use the Pi4J library and create my own Project? And is out there any Pin diagram with all BCM infos for the pins?

**Comment by FDelporte** [comment]
That JBang is only an example! Ofcourse you can use your own project, or Maven, or Gradle, ...

On [https://pinout.xyz/pinout](https://pinout.xyz/pinout/pin40_gpio21/#) you can find all pin info.

**Reply by quodix** [reply]
One more question.
I want to open Serial port with TX and RX function.
How can I set the RX and TX in the SerialConfig? I only have "Device" to set

**Reply by FDelporte** [reply]
See https://pi4j.com/documentation/io-examples/serial/

.device("/dev/ttyS0") uses 

RX | UART TX, GPIO 15
TX | UART RX, GPIO 16
